keepsorted/line_group.go

```diff
@@ -15,25 +15,95 @@
 package keepsorted
 
 import (
-	"cmp"
 	"fmt"
 	"regexp"
 	"strings"
+	"sync"
 	"unicode"
 
 	"github.com/rs/zerolog/log"
 )
 
-// lineGroup is a logical unit of source code. It's one or more lines combines
+// lineGroup is a logical unit of source code. It's one or more lines combined
 // with zero or more comment lines about the source code lines.
 type lineGroup struct {
+	opts        blockOptions
+	prefixOrder func() *prefixOrder
+
+	// The actual content of the lineGroup.
+	lineGroupContent
+
+	// Bits of information that we use while comparing two line groups together.
+	// These bits of information are calculated lazily, and are invalidated when
+	// the lineGroup is mutated. You should prefer to access them via their
+	// getters.
+	calculated lineGroupCalculations
+}
+
+var compareLineGroups = comparingFunc((*lineGroup).commentOnly, falseFirst()).
+	andThen(comparingFunc((*lineGroup).regexTokens, lexicographically(compareRegexTokens))).
+	andThen(comparing((*lineGroup).joinedLines)).
+	andThen(comparing((*lineGroup).joinedComment))
+
+var compareRegexTokens = comparingFunc(func(t regexToken) bool { return t == nil }, falseFirst()).
+	andThen(comparingFunc(func(t regexToken) []*captureGroupToken { return t }, lexicographically(compareCaptureGroupTokens)))
+
+var compareCaptureGroupTokens = comparingFunc((*captureGroupToken).prefix, orderedPrefix.compare).
+	andThen(comparingFunc((*captureGroupToken).transform, numericTokens.compare))
+
+type lineGroupContent struct {
 	comment []string
 	lines   []string
 }
 
+type lineGroupCalculations struct {
+	commentOnly *bool
+	regexes     []regexToken
+	lines       string
+	comment     string
+}
+
+type regexToken []*captureGroupToken
+
+type captureGroupToken struct {
+	opts        blockOptions
```

Comment:
Should this be here? It's part of the lineGroup, so perhaps could be accessed that way instead (either by a back reference to the lineGroup or by changing signatures so that the lineGroup is a parameter). Or at the very least, perhaps this should become a pointer type (the other options are a smidge more awkward). After all, it's growing quite large. It'll be the same for an entire list of lineGroups (i.e., each keep-sorted block), so it seems awkward to be storing many copies of it within the block.

Reply:
Hmm I'm not sure how much of a concern this is. blockOptions should be pretty cheap to copy, no? blockOptions only has 15 fields, and the slices and maps would only get a shallow copy (right?)

I don't have a great sense for when this would be a concern, but I'm happy enough to make this field a pointer to avoid unnecessary copies